File: packages/sdk-python/src/acruxcore/evaluations.py
"""Evaluations namespaces: datasets, experiments, runs, optimize."""

from __future__ import annotations

from typing import TYPE_CHECKING, Any, Dict, List, Optional
from urllib.parse import quote

if TYPE_CHECKING:
    from .client import AcruxCore

from .eval_types import (
    BuildFromFeedbackResult,
    CandidateDetail,
    DatasetDto,
    DatasetExampleDto,
    DatasetWithExamples,
    ExperimentDto,
    PromoteResult,
    RunCellDetailDto,
    RunDetailDto,
    RunListResponse,
    RunReport,
    StartOptimizeResult,
    StartRunResult,
)


class DatasetsNamespace:
    """``client.datasets`` — create, list, get, update, and delete datasets."""

    def __init__(self, client: "AcruxCore") -> None:
        self._client = client

    async def create(
        self,
        name: str,
        *,
        overall_feedback: Optional[str] = None,
    ) -> DatasetDto:
        body: Dict[str, Any] = {"name": name}
        if overall_feedback is not None:
            body["overall_feedback"] = overall_feedback
        response = await self._client._request(
            "POST", "/datasets", body, "creating dataset"
        )
        return DatasetDto.from_dict(
            self._client._parse_json_or_throw(response, "creating dataset")
        )

    async def build_from_feedback(
        self,
        name: str,
        feedback_ids: List[str],
        *,
        overall_feedback: Optional[str] = None,
    ) -> BuildFromFeedbackResult:
        body: Dict[str, Any] = {
            "name": name,
            "feedback_ids": feedback_ids,
        }
        if overall_feedback is not None:
            body["overall_feedback"] = overall_feedback
        response = await self._client._request(
            "POST", "/datasets/from-feedback", body, "building dataset from feedback"
        )
        return BuildFromFeedbackResult.from_dict(
            self._client._parse_json_or_throw(
                response, "building dataset from feedback"
            )
        )

    async def list(self) -> List[DatasetDto]:
        response = await self._client._request(
            "GET", "/datasets", None, "listing datasets"
        )
        data = self._client._parse_json_or_throw(response, "listing datasets")
        return [DatasetDto.from_dict(item) for item in (data.get("data") or [])]

    async def get(self, dataset_id: str) -> DatasetWithExamples:
        response = await self._client._request(
            "GET",
            f"/datasets/{quote(dataset_id, safe='')}",
            None,
            "getting dataset",
        )
        return DatasetWithExamples.from_dict(
            self._client._parse_json_or_throw(response, "getting dataset")
        )

    async def update(
        self,
        dataset_id: str,
        **kwargs: Any,
    ) -> DatasetDto:
        body: Dict[str, Any] = {}
        if "name" in kwargs:
            body["name"] = kwargs["name"]
        if "overall_feedback" in kwargs:
            body["overall_feedback"] = kwargs["overall_feedback"]
        response = await self._client._request(
            "PATCH",
            f"/datasets/{quote(dataset_id, safe='')}",
            body,
            "updating dataset",
        )
        return DatasetDto.from_dict(
            self._client._parse_json_or_throw(response, "updating dataset")
        )

    async def delete(self, dataset_id: str) -> None:
        await self._client._request(
            "DELETE",
            f"/datasets/{quote(dataset_id, safe='')}",
            None,
            "deleting dataset",
        )

    async def add_example(
        self,
        dataset_id: str,
        input: Dict[str, Any],
        *,
        criteria: Optional[str] = None,
        history: Optional[List[Dict[str, Any]]] = None,
    ) -> DatasetExampleDto:
        body: Dict[str, Any] = {"input": input}
        if criteria is not None:
            body["criteria"] = criteria
        if history is not None:
            body["history"] = history
        response = await self._client._request(
            "POST",
            f"/datasets/{quote(dataset_id, safe='')}/examples",
            body,
            "adding example",
        )
        return DatasetExampleDto.from_dict(
            self._client._parse_json_or_throw(response, "adding example")
        )

    async def remove_example(
        self, dataset_id: str, example_id: str
    ) -> None:
        await self._client._request(
            "DELETE",
            f"/datasets/{quote(dataset_id, safe='')}/examples/{quote(example_id, safe='')}",
            None,
            "removing example",
        )


class ExperimentsNamespace:
    """``client.experiments`` — create, list, get experiments and start runs."""

    def __init__(self, client: "AcruxCore") -> None:
        self._client = client

    async def create(
        self,
        dataset_id: str,
        version_ids: List[str],
        models: List[str],
        *,
        prompt_id: Optional[str] = None,
        name: Optional[str] = None,
        alias: Optional[str] = None,
    ) -> ExperimentDto:
        body: Dict[str, Any] = {
            "dataset_id": dataset_id,
            "version_ids": version_ids,
            "models": models,
        }
        if prompt_id is not None:
            body["prompt_id"] = prompt_id
        if name is not None:
            body["name"] = name
        if alias is not None:
            body["alias"] = alias
        response = await self._client._request(
            "POST", "/experiments", body, "creating experiment"
        )
        return ExperimentDto.from_dict(
            self._client._parse_json_or_throw(response, "creating experiment")
        )

    async def list(self) -> List[ExperimentDto]:
        response = await self._client._request(
            "GET", "/experiments", None, "listing experiments"
        )
        data = self._client._parse_json_or_throw(response, "listing experiments")
        return [ExperimentDto.from_dict(item) for item in (data.get("data") or [])]

    async def get(self, experiment_id: str) -> ExperimentDto:
        response = await self._client._request(
            "GET",
            f"/experiments/{quote(experiment_id, safe='')}",
            None,
            "getting experiment",
        )
        return ExperimentDto.from_dict(
            self._client._parse_json_or_throw(response, "getting experiment")
        )

    async def start_run(self, experiment_id: str) -> StartRunResult:
        response = await self._client._request(
            "POST",
            f"/experiments/{quote(experiment_id, safe='')}/runs",
            None,
            "starting run",
        )
        return StartRunResult.from_dict(
            self._client._parse_json_or_throw(response, "starting run")
        )


class RunsNamespace:
    """``client.runs`` — list, get, read reports, inspect cells, promote candidates."""

    def __init__(self, client: "AcruxCore") -> None:
        self._client = client

    async def list(
        self,
        *,
        status: Optional[str] = None,
        dataset_id: Optional[str] = None,
        prompt_id: Optional[str] = None,
        page: Optional[int] = None,
        limit: Optional[int] = None,
    ) -> RunListResponse:
        params: Dict[str, str] = {}
        if status:
            params["status"] = status
        if dataset_id:
            params["dataset_id"] = dataset_id
        if prompt_id:
            params["prompt_id"] = prompt_id
        if page is not None:
            params["page"] = str(page)
        if limit is not None:
            params["limit"] = str(limit)
        qs = "&".join(f"{k}={v}" for k, v in params.items()) if params else ""
        path = f"/runs?{qs}" if qs else "/runs"
        response = await self._client._request(
            "GET", path, None, "listing runs"
        )
        return RunListResponse.from_dict(
            self._client._parse_json_or_throw(response, "listing runs")
        )

    async def get(self, run_id: str) -> RunDetailDto:
        response = await self._client._request(
            "GET",
            f"/runs/{quote(run_id, safe='')}",
            None,
            "getting run",
        )
        return RunDetailDto.from_dict(
            self._client._parse_json_or_throw(response, "getting run")
        )

    async def get_report(self, run_id: str) -> RunReport:
        response = await self._client._request(
            "GET",
            f"/runs/{quote(run_id, safe='')}/report",
            None,
            "getting run report",
        )
        return RunReport.from_dict(
            self._client._parse_json_or_throw(response, "getting run report")
        )

    async def get_cell(self, run_id: str, cell_key: str) -> RunCellDetailDto:
        response = await self._client._request(
            "GET",
            f"/runs/{quote(run_id, safe='')}/cells/{quote(cell_key, safe='')}",
            None,
            "getting run cell",
        )
        return RunCellDetailDto.from_dict(
            self._client._parse_json_or_throw(response, "getting run cell")
        )

    async def get_candidate(
        self, run_id: str, candidate_id: str
    ) -> CandidateDetail:
        response = await self._client._request(
            "GET",
            f"/runs/{quote(run_id, safe='')}/candidates/{quote(candidate_id, safe='')}",
            None,
            "getting candidate",
        )
        return CandidateDetail.from_dict(
            self._client._parse_json_or_throw(response, "getting candidate")
        )

    async def promote_candidate(
        self,
        run_id: str,
        prompt_candidate_id: str,
        *,
        alias: Optional[str] = None,
    ) -> PromoteResult:
        body: Dict[str, Any] = {"prompt_candidate_id": prompt_candidate_id}
        if alias is not None:
            body["alias"] = alias
        response = await self._client._request(
            "POST",
            f"/runs/{quote(run_id, safe='')}/promote",
            body,
            "promoting candidate",
        )
        return PromoteResult.from_dict(
            self._client._parse_json_or_throw(response, "promoting candidate")
        )


class OptimizeNamespace:
    """``client.optimize`` — start prompt optimization runs."""

    def __init__(self, client: "AcruxCore") -> None:
        self._client = client

    async def start(
        self,
        prompt_id: str,
        dataset_id: str,
        models: List[str],
        *,
        draft_count: Optional[int] = None,
        alias: Optional[str] = None,
    ) -> StartOptimizeResult:
        body: Dict[str, Any] = {
            "dataset_id": dataset_id,
            "models": models,
        }
        if draft_count is not None:
            body["draft_count"] = draft_count
        if alias is not None:
            body["alias"] = alias
        response = await self._client._request(
            "POST",
            f"/prompts/{quote(prompt_id, safe='')}/optimize",
            body,
            "starting optimize",
        )
        return StartOptimizeResult.from_dict(
            self._client._parse_json_or_throw(response, "starting optimize")
        )

File: packages/sdk/src/evaluations.ts
import { acruxcoreError } from './error';
import type {
  CreateDatasetParams,
  BuildFromFeedbackParams,
  BuildFromFeedbackResult,
  UpdateDatasetParams,
  AddExampleParams,
  DatasetDto,
  DatasetWithExamples,
  DatasetExampleDto,
  CreateExperimentParams,
  ExperimentDto,
  ExperimentRunDto,
  StartRunResult,
  ListRunsParams,
  RunListResponse,
  RunDetailDto,
  RunReport,
  RunCellDetailDto,
  CandidateDetail,
  PromoteCandidateParams,
  PromoteResult,
  StartOptimizeParams,
  StartOptimizeResult,
} from './eval-types';
import type { NamespaceHost } from './host';

export type EvaluationsHost = NamespaceHost;

// ── Datasets ──

export class DatasetsNamespace {
  private readonly client: EvaluationsHost;

  constructor(client: EvaluationsHost) {
    this.client = client;
  }

  async create(params: CreateDatasetParams): Promise<DatasetDto> {
    const response = await this.client._request('POST', '/datasets', params as unknown as Record<string, unknown>, 'creating dataset');
    return this.client._parseJsonOrThrow(response, 'creating dataset') as Promise<DatasetDto>;
  }

  async buildFromFeedback(params: BuildFromFeedbackParams): Promise<BuildFromFeedbackResult> {
    const body = {
      name: params.name,
      ...(params.overallFeedback !== undefined ? { overall_feedback: params.overallFeedback } : {}),
      feedback_ids: params.feedbackIds,
    };
    const response = await this.client._request('POST', '/datasets/from-feedback', body, 'building dataset from feedback');
    return this.client._parseJsonOrThrow(response, 'building dataset from feedback') as Promise<BuildFromFeedbackResult>;
  }

  async list(): Promise<DatasetDto[]> {
    const response = await this.client._request('GET', '/datasets', undefined, 'listing datasets');
    const data = await this.client._parseJsonOrThrow(response, 'listing datasets') as { data?: DatasetDto[] };
    return data.data ?? [];
  }

  async get(id: string): Promise<DatasetWithExamples> {
    const response = await this.client._request('GET', `/datasets/${encodeURIComponent(id)}`, undefined, 'getting dataset');
    return this.client._parseJsonOrThrow(response, 'getting dataset') as Promise<DatasetWithExamples>;
  }

  async update(id: string, params: UpdateDatasetParams): Promise<DatasetDto> {
    const body: Record<string, unknown> = {};
    if (params.name !== undefined) body['name'] = params.name;
    if (params.overallFeedback !== undefined) body['overall_feedback'] = params.overallFeedback;
    const response = await this.client._request('PATCH', `/datasets/${encodeURIComponent(id)}`, body, 'updating dataset');
    return this.client._parseJsonOrThrow(response, 'updating dataset') as Promise<DatasetDto>;
  }

  async delete(id: string): Promise<{ success: boolean }> {
    const response = await this.client._request('DELETE', `/datasets/${encodeURIComponent(id)}`, undefined, 'deleting dataset');
    return this.client._parseJsonOrThrow(response, 'deleting dataset') as Promise<{ success: boolean }>;
  }

  async addExample(datasetId: string, params: AddExampleParams): Promise<DatasetExampleDto> {
    const response = await this.client._request('POST', `/datasets/${encodeURIComponent(datasetId)}/examples`, params as unknown as Record<string, unknown>, 'adding example');
    return this.client._parseJsonOrThrow(response, 'adding example') as Promise<DatasetExampleDto>;
  }

  async removeExample(datasetId: string, exampleId: string): Promise<{ success: boolean }> {
    const response = await this.client._request('DELETE', `/datasets/${encodeURIComponent(datasetId)}/examples/${encodeURIComponent(exampleId)}`, undefined, 'removing example');
    return this.client._parseJsonOrThrow(response, 'removing example') as Promise<{ success: boolean }>;
  }
}

// ── Experiments ──

export class ExperimentsNamespace {
  private readonly client: EvaluationsHost;

  constructor(client: EvaluationsHost) {
    this.client = client;
  }

  async create(params: CreateExperimentParams): Promise<ExperimentDto> {
    const body = {
      dataset_id: params.datasetId,
      ...(params.promptId !== undefined ? { prompt_id: params.promptId } : {}),
      ...(params.name !== undefined ? { name: params.name } : {}),
      version_ids: params.versionIds,
      models: params.models,
      ...(params.alias !== undefined ? { alias: params.alias } : {}),
    };
    const response = await this.client._request('POST', '/experiments', body, 'creating experiment');
    return this.client._parseJsonOrThrow(response, 'creating experiment') as Promise<ExperimentDto>;
  }

  async list(): Promise<ExperimentDto[]> {
    const response = await this.client._request('GET', '/experiments', undefined, 'listing experiments');
    const data = await this.client._parseJsonOrThrow(response, 'listing experiments') as { data?: ExperimentDto[] };
    return data.data ?? [];
  }

  async get(id: string): Promise<ExperimentDto> {
    const response = await this.client._request('GET', `/experiments/${encodeURIComponent(id)}`, undefined, 'getting experiment');
    return this.client._parseJsonOrThrow(response, 'getting experiment') as Promise<ExperimentDto>;
  }

  async startRun(experimentId: string): Promise<StartRunResult> {
    const response = await this.client._request('POST', `/experiments/${encodeURIComponent(experimentId)}/runs`, undefined, 'starting run');
    const data = await this.client._parseJsonOrThrow(response, 'starting run') as { run_id: string; status: string };
    return { runId: data.run_id, status: data.status };
  }
}

// ── Runs ──

export class RunsNamespace {
  private readonly client: EvaluationsHost;

  constructor(client: EvaluationsHost) {
    this.client = client;
  }

  async list(params: ListRunsParams = {}): Promise<RunListResponse> {
    const qs = new URLSearchParams();
    if (params.status) qs.set('status', params.status);
    if (params.datasetId) qs.set('dataset_id', params.datasetId);
    if (params.promptId) qs.set('prompt_id', params.promptId);
    if (params.page !== undefined) qs.set('page', String(params.page));
    if (params.limit !== undefined) qs.set('limit', String(params.limit));
    const query = qs.toString();
    const path = `/runs${query ? `?${query}` : ''}`;
    const response = await this.client._request('GET', path, undefined, 'listing runs');
    return this.client._parseJsonOrThrow(response, 'listing runs') as Promise<RunListResponse>;
  }

  async get(id: string): Promise<RunDetailDto> {
    const response = await this.client._request('GET', `/runs/${encodeURIComponent(id)}`, undefined, 'getting run');
    return this.client._parseJsonOrThrow(response, 'getting run') as Promise<RunDetailDto>;
  }

  async getReport(id: string): Promise<RunReport> {
    const response = await this.client._request('GET', `/runs/${encodeURIComponent(id)}/report`, undefined, 'getting run report');
    return this.client._parseJsonOrThrow(response, 'getting run report') as Promise<RunReport>;
  }

  async getCell(id: string, cellKey: string): Promise<RunCellDetailDto> {
    const response = await this.client._request('GET', `/runs/${encodeURIComponent(id)}/cells/${encodeURIComponent(cellKey)}`, undefined, 'getting run cell');
    return this.client._parseJsonOrThrow(response, 'getting run cell') as Promise<RunCellDetailDto>;
  }

  async getCandidate(id: string, candidateId: string): Promise<CandidateDetail> {
    const response = await this.client._request('GET', `/runs/${encodeURIComponent(id)}/candidates/${encodeURIComponent(candidateId)}`, undefined, 'getting candidate');
    return this.client._parseJsonOrThrow(response, 'getting candidate') as Promise<CandidateDetail>;
  }

  async promoteCandidate(id: string, params: PromoteCandidateParams): Promise<PromoteResult> {
    const body: Record<string, unknown> = { prompt_candidate_id: params.promptCandidateId };
    if (params.alias !== undefined) body['alias'] = params.alias;
    const response = await this.client._request('POST', `/runs/${encodeURIComponent(id)}/promote`, body, 'promoting candidate');
    return this.client._parseJsonOrThrow(response, 'promoting candidate') as Promise<PromoteResult>;
  }
}

// ── Optimize ──

export class OptimizeNamespace {
  private readonly client: EvaluationsHost;

  constructor(client: EvaluationsHost) {
    this.client = client;
  }

  async start(promptId: string, params: StartOptimizeParams): Promise<StartOptimizeResult> {
    const body: Record<string, unknown> = {
      dataset_id: params.datasetId,
      models: params.models,
    };
    if (params.draftCount !== undefined) body['draft_count'] = params.draftCount;
    if (params.alias !== undefined) body['alias'] = params.alias;
    const response = await this.client._request('POST', `/prompts/${encodeURIComponent(promptId)}/optimize`, body, 'starting optimize');
    const data = await this.client._parseJsonOrThrow(response, 'starting optimize') as { run_id: string; status: string; prompt_mismatch_warning?: StartOptimizeResult['promptMismatchWarning'] };
    return { runId: data.run_id, status: data.status, promptMismatchWarning: data.prompt_mismatch_warning };
  }
}
